mllib/src/main/scala/org/apache/spark/ml/feature/RModelFormula.scala

```diff
@@ -0,0 +1,121 @@
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types._
+
+/**
+ * :: Experimental ::
+ * Implements the transforms required for fitting a dataset against a R model formula.
+ */
+@Experimental
+private[spark] class RModelFormula(override val uid: String)
+  extends Transformer with HasFeaturesCol with HasLabelCol {
+
+  def this() = this(Identifiable.randomUID("rModelFormula"))
+
+  val formula: Param[String] = new Param(this, "formula", "R model formula")
+  protected var parsedFormula: Option[RFormula] = None
+
+  /**
+   * Sets the formula to use for this transformer. Must be called before use.
+   * @param value a R formula in string form (e.g. "y ~ x + z")
+   */
+  def setFormula(value: String): this.type = {
+    parsedFormula = Some(RFormulaParser.parse(value))
+    set(formula, value)
+    this
+  }
+
+  override def transformSchema(schema: StructType): StructType = {
+    require(parsedFormula.isDefined, "Must call setFormula() first.")
+    val withFeatures = featureTransformer.transformSchema(schema)
+    val nullable = schema(parsedFormula.get.response).dataType match {
+      case _: NumericType | BooleanType => false
+      case _ => true
+    }
+    StructType(withFeatures.fields :+ StructField($(labelCol), DoubleType, nullable))
+  }
+
+  override def transform(dataset: DataFrame): DataFrame = {
+    require(parsedFormula.isDefined, "Must call setFormula() first.")
+    transformLabel(featureTransformer.transform(dataset))
+  }
+
+  override def copy(extra: ParamMap): RModelFormula = defaultCopy(extra)
+
+  override def toString: String = s"RModelFormula(${get(formula)})"
+
+  protected def transformLabel(dataset: DataFrame): DataFrame = {
+    val responseName = parsedFormula.get.response
+    dataset.schema(responseName).dataType match {
+      case _: NumericType | BooleanType =>
+        dataset.select(
+          col("*"),
+          dataset(responseName).cast(DoubleType).as($(labelCol)))
+      case StringType =>
+        new StringIndexer(uid)
+          .setInputCol(responseName)
+          .setOutputCol($(labelCol))
+          .fit(dataset)
+          .transform(dataset)
+      case other =>
+        throw new IllegalArgumentException("Unsupported type for response: " + other)
+    }
+  }
+
+  protected def featureTransformer: Transformer = {
+    // TODO(ekl) add support for non-numeric features and feature interactions
+    new VectorAssembler(uid)
+      .setInputCols(parsedFormula.get.terms.toArray)
+      .setOutputCol($(featuresCol))
+  }
+}
+
+/**
+ * :: Experimental ::
+ * Represents a parsed R formula.
+ */
+private[ml] case class RFormula(response: String, terms: Seq[String])
+
+/**
+ * :: Experimental ::
+ * Limited implementation of R formula parsing. Currently supports: '~', '+'.
+ */
+private[ml] object RFormulaParser extends RegexParsers {
+  def term: Parser[String] = "([a-zA-Z]|\\.[a-zA-Z_])[a-zA-Z0-9._]*".r
```

Comment:
Looks like R supports arbitrary expressions in terms, so we'd need a full parser to be sure. For $ am I not sure it makes sense, since we assume the terms are from the dataframe already.